[[ Bug 19783 ]] Clear errors at end of try

montegoulding opened the pull request:
At the end of a try control structure any errors on the error
stack should be cleared. This was not happening if an error was
added to the error stack from some other context. Until we have
more context sensitive try locks and error stacks we should do
our best to clear any unthrown errors on the stack.

runrevmark commented:
@montegoulding Is there a bug reported about improving the way 'try' works?

@livecode-vulcan review ok f72fa88

livecode-vulcan commented:
💙 review by @runrevmark ok f72fa884cf5f94670f16a277296cd706f3c73bae

livecodepanos commented on docs/notes/bugfix-19783.md:
```diff
@@ -0,0 +1 @@
+# Ensure any errors are cleared from teh error stack after a try control structure
```
There is a typo here (`teh`) :)

livecode-vulcan commented:
😎 test success f72fa884cf5f94670f16a277296cd706f3c73bae

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/336)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/342)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/322)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/265)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/265)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/331)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/336)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/334)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/340)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/334)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/340)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/359)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/330)

montegoulding commented:
@runrevmark not yet. Now that this has been accepted I'll move the rest of bug 19783 to a moomin report. If this wasn't accepted I would have turned 19783 into a moomin ;-)

montegoulding commented:
http://quality.livecode.com/show_bug.cgi?id=19805

montegoulding committed:
[[ Bug 19783 ]] Don't retain errors from other events in try

This patch ensures that if statements are executed successfully that
there are no errors on the error stack. This may happen if the error
was from a different event and not thrown due to the trylock.

montegoulding committed:
[[ Bug 19783 ]] Ensure all error execution states are handled by try

This patch adds not handled and not found as possible error states for
try errors.

montegoulding committed:
[[ Bug 19783 ]] Ensure try execs finally

This patch ensures that even if the try block has no statements that
the finally block will be executed.

montegoulding committed:
[[ Bug 19783 ]] Ensure errors in catch and finally are thrown

This patch ensures that the context has it's exec state set
back if there was an error in the catch or finally statements

montegoulding committed:
[[ Bug 19783 ]] Ensure errors are thrown in finally block

This patch ensures errors are not cleared if they occur in a finally
block. Previously they were cleared and the exec state wa forced to
normal.

montegoulding committed:
[[ Bug 19810 ]] Add pass error form of pass

This patch adds `pass error` as an option to the `pass` command. This
is required because at present the syntax requiring `pass handlerName`
is ambiguous with respect to try.

This patch also ensures that the mutation of the error variable
does not mutate the error that is passed.

montegoulding committed:
[[ Bug 13482 ]] Document optional catch clause in try

montegoulding committed:
[[ Tests ]] Add tests for try control structure

montegoulding commented:
@runrevmark I will need to drop 4daafa3 and related tests from this patch as it adds a feature. I would be pleased to have general feedback before I do anything else here though. Can you think of anything else that needs testing?

runrevmark commented:
I suggest splitting this up as follows (assuming it is possible):

- The patch which fixes the issue we encountered in the test runner (develop)
- The patch which fixes the docs about try being optional (develop-8.1)
- The patch which has the tests, with the tests which are broken marked as TestBroked (develop) - this means we can then iterate on them to make sure they represent the functionality we want (as there are few edge cases with regards 'return' and 'pass <handler>' as you've pointed out).
- A (WIP) patch which fixes the tests as current written.

If we then decide that some of the changes are too deep for 9 we can hold over until the next version.

montegoulding commented:
OK, sure @runrevmark will do
